File: leetcode/1518.py
class Solution:
    def numWaterBottles(self, numBottles: int, numExchange: int) -> int:
        def loop(empties: int, drunks: int) -> int:
            if empties >= numExchange:
                return loop(empties - numExchange + 1, drunks + 1)
            return drunks

        return loop(numBottles, numBottles)

File: leetcode/1518.cpp
class Solution {
  public:
    int numWaterBottles(int numBottles, int numExchange) {
        auto loop = [numExchange](this auto &&self, int empties,
                                  int drunks) -> int {
            if (empties >= numExchange)
                return self(empties - numExchange + 1, drunks + 1);

            return drunks;
        };

        return loop(numBottles, numBottles);
    }
};
